### PR TITLE
add a list handling to manage multiple users or groups with same perm…

### PR DESCRIPTION
**Pull request by robert8103**
**What does this PR do?**
Handle ACL lists and added some documentation.
I also added some more documentation.

**Previous Behavior**
You are only able to add acl with only one acl_name

**New Behavior**
Now you are also able to handle multiple acl_names in lists with the same rights

```yaml
root:
  acl.list_present:
    - name: /root
    - acl_type: user
    - acl_name:
      - damian
      - homer
    - perms: rwx
###
root:
  acl.list_absent:
    - name: /root
    - acl_type: user
    - acl_name:
      - damian
      - homer
    - perms: rwx
```

**Tests written?**

No
**Commits signed with GPG?**

No

### CODE REVIEW

**Comment by garethgreenaway** [comment]
@robert8103 Looks good.  Would you mind adding some tests for this new functionality?

**Comment by robert8103** [comment]
@garethgreenaway obviously yes, but i'm not very experienced in writing python tests.
maybe you can give me some hints to write an appropriate test for this.

File: salt/states/linux_acl.py [hunk 1]
```diff
@@ -221,3 +284,241 @@ def absent(name, acl_type, acl_name='', perms='', recurse=False):
...
+    # We search through the dictionary getfacl returns for the owner of the
+    # file if acl_name is empty.
+    if acl_names == '':
+        _search_name = __current_perms[name].get('comment').get(_acl_type, '')
+    else:
+        _search_name = acl_names
+
+    if _current_perms.get(_acl_type, None) or _default:
+        try:
+            user = [i for i in _current_perms[_acl_type] if next(six.iterkeys(i)) == _search_name].pop()
```
Comment (garethgreenaway) [inline salt/states/linux_acl.py]: Based on the code at lines 367 - 370 I don't see how they would ever match.  The `acl_names` would be a list and in turn _search_name would be a list, but the code in the list comprehension is going to produce a single value.

**Comment by garethgreenaway** [comment]
@robert8103 I started an effort to add some unit tests for this change and keep running into situations where the results being returned aren't the expected results.  When I run it with a test state file it's always trying to set the ACLs.  As I indicated in my line comment, I think the logic for grabbing the `user` and `_search_name` isn't working correctly.

**Comment by robert8103** [comment]
@garethgreenaway I added now the handling with the list.
may be you add your tests that i can use them if you run into more bugs.

**Comment by rallytime** [comment]
@robert8103 Looks like there is a merge conflict on this change now. Can you rebase this and resolve the conflicts?

**Comment by rallytime** [comment]
Hi @robert8103 - This is still conflicting. Please let us know if you have questions!

**Comment by robert8103** [comment]
hi @rallytime - yes i have little problem to solve this conflicts.
it would be grad if you could help me.

**Comment by robert8103** [comment]
Hi @rallytime, thank you for your help!
Now it should be up to date.

**Comment by cachedout** [comment]
I'm quite nervous about merging this without tests. While I appreciate the effort here, if we can't get tests submitted with this I don't think we can accept it.

**Comment by gtmanfred** [comment]
There is a #testing and #develop channel on the slack listed here
https://github.com/saltstack/salt/#get-saltstack-support-and-help

**Review by gtmanfred — COMMENTED** [review]
Still just waiting on tests.

**Comment by garethgreenaway** [comment]
@robert8103 I spent some time looking at this over the weekend and added some initial tests based on the existing `present` and `absent` ones.  I had to tweak the code a bit to account for some previous unaccounted situations.
```
diff --git a/salt/states/linux_acl.py b/salt/states/linux_acl.py
index 6834456aee..fa52b7f766 100644
--- a/salt/states/linux_acl.py
+++ b/salt/states/linux_acl.py
@@ -341,9 +341,11 @@ def list_present(name, acl_type, acl_names=None, perms='', recurse=False, force=
                 _current_acl_types.append(current_acl_name.encode('utf-8'))
                 diff_perms = _octal_perms == key[current_acl_name]['octal']
         if acl_type == 'user':
-            _current_acl_types.remove(_origin_owner)
+            if _origin_owner:
+                _current_acl_types.remove(_origin_owner)
         else:
-            _current_acl_types.remove(_origin_group)
+            if _origin_group:
+                _current_acl_types.remove(_origin_group)
         diff_acls = set(_current_acl_types) ^ set(acl_names)
         if not diff_acls and diff_perms and not force:
             ret = {'name': name,
@@ -367,30 +369,60 @@ def list_present(name, acl_type, acl_names=None, perms='', recurse=False, force=
 
     if _current_perms.get(_acl_type, None) or _default:
         try:
-            users = [i for i in _current_perms[_acl_type] if next(six.iterkeys(i)) in _search_names].pop()
+            users = {}
+            for i in _current_perms[_acl_type]:
+                if i and next(six.iterkeys(i)) in _search_names:
+                    users.update(i)
         except (AttributeError, KeyError):
             users = None
 
         if users:
             changes = {}
             for count, search_name in enumerate(_search_names):
-                if users[search_name]['octal'] == sum([_octal.get(i, i) for i in perms]):
-                    ret['comment'] = 'Permissions are in the desired state'
+                if search_name in users:
+                    if users[search_name]['octal'] == sum([_octal.get(i, i) for i in perms]):
+                        ret['comment'] = 'Permissions are in the desired state'
+                    else:
+                        changes.update({'new': {'acl_name': ', '.join(acl_names),
+                                                                 'acl_type': acl_type,
+                                                                 'perms': _octal_perms},
+                                        'old': {'acl_name': ', '.join(acl_names),
+                                                                 'acl_type': acl_type,
+                                                                 'perms': six.text_type(users[search_name]['octal'])}})
+                        if __opts__['test']:
+                            ret.update({'comment': 'Updated permissions will be applied for '
+                                                   '{0}: {1} -> {2}'.format(
+                                acl_names,
+                                six.text_type(users[search_name]['octal']),
+                                perms),
+                                'result': None, 'pchanges': changes})
+                            return ret
+                        try:
+                            if force:
+                                __salt__['acl.wipefacls'](name, recursive=recurse, raise_err=True)
+
+                            for acl_name in acl_names:
+                                __salt__['acl.modfacl'](acl_type, acl_name, perms, name,
+                                                        recursive=recurse, raise_err=True)
+                            ret.update({'comment': 'Updated permissions for '
+                                                   '{0}'.format(acl_names),
+                                        'result': True, 'changes': changes})
+                        except CommandExecutionError as exc:
+                            ret.update({'comment': 'Error updating permissions for '
+                                                   '{0}: {1}'.format(acl_names, exc.strerror),
+                                        'result': False})
                 else:
-                    changes.update({'new': {'acl_name': ', '.join(acl_names),
-                                                             'acl_type': acl_type,
-                                                             'perms': _octal_perms},
-                                    'old': {'acl_name': ', '.join(acl_names),
-                                                             'acl_type': acl_type,
-                                                             'perms': six.text_type(users[search_name]['octal'])}})
+                    changes = {'new': {'acl_name': ', '.join(acl_names),
+                                       'acl_type': acl_type,
+                                       'perms': perms}}
+
                     if __opts__['test']:
-                        ret.update({'comment': 'Updated permissions will be applied for '
-                                               '{0}: {1} -> {2}'.format(
-                            acl_names,
-                            six.text_type(users[_search_names]['octal']),
-                            perms),
-                            'result': None, 'pchanges': changes})
+                        ret.update({'comment': 'New permissions will be applied for '
+                                               '{0}: {1}'.format(acl_names, perms),
+                                    'result': None, 'pchanges': changes})
+                        ret['result'] = None
                         return ret
+
                     try:
                         if force:
                             __salt__['acl.wipefacls'](name, recursive=recurse, raise_err=True)
@@ -398,15 +430,16 @@ def list_present(name, acl_type, acl_names=None, perms='', recurse=False, force=
                         for acl_name in acl_names:
                             __salt__['acl.modfacl'](acl_type, acl_name, perms, name,
                                                     recursive=recurse, raise_err=True)
-                        ret.update({'comment': 'Updated permissions for '
-                                               '{0}'.format(acl_names),
+                        ret.update({'comment': 'Applied new permissions for '
+                                               '{0}'.format(', '.join(acl_names)),
                                     'result': True, 'changes': changes})
                     except CommandExecutionError as exc:
-                        ret.update({'comment': 'Error updating permissions for '
-                                               '{0}: {1}'.format(acl_names, exc.strerror),
+                        ret.update({'comment': 'Error updating permissions for {0}: '
+                                               '{1}'.format(acl_names, exc.strerror),
                                     'result': False})
+
         else:
-            changes = {'new': {'acl_name': acl_names,
+            changes = {'new': {'acl_name': ', '.join(acl_names),
                                'acl_type': acl_type,
                                'perms': perms}}
 
@@ -426,7 +459,7 @@ def list_present(name, acl_type, acl_names=None, perms='', recurse=False, force=
                                             recursive=recurse, raise_err=True)
                 ret.update({'comment': 'Applied new permissions for '
                                        '{0}'.format(', '.join(acl_names)),
-                            'result': True, 'changes': {}})
+                            'result': True, 'changes': changes})
             except CommandExecutionError as exc:
                 ret.update({'comment': 'Error updating permissions for {0}: '
                                        '{1}'.format(acl_names, exc.strerror),
diff --git a/tests/unit/states/test_linux_acl.py b/tests/unit/states/test_linux_acl.py
index 2b31b45297..8554e4de4f 100644
--- a/tests/unit/states/test_linux_acl.py
+++ b/tests/unit/states/test_linux_acl.py
@@ -180,3 +180,170 @@ class LinuxAclTestCase(TestCase, LoaderModuleMockMixin):
             comt = ('ACL Type does not exist')
             ret.update({'comment': comt, 'result': False})
             self.assertDictEqual(linux_acl.absent(name, acl_type, acl_name, perms), ret)
+
+    # 'list_present' function tests: 1
+
+    def test_list_present(self):
+        '''
+        Test to ensure a Linux ACL is present
+        '''
+        self.maxDiff = None
+        name = '/root'
+        acl_type = 'user'
+        acl_names = ['root', 'damian', 'homer']
+        acl_comment = {u'owner': u'root',
+                       u'group': u'root',
+                       u'file': u'/root'}
+        perms = 'rwx'
+
+        mock = MagicMock(side_effect=[{name: {acl_type: [{acl_names[0]:
+                                                         {'octal': 'A'}},
+                                                         {acl_names[1]:
+                                                         {'octal': 'A'}},
+                                                         {acl_names[2]:
+                                                         {'octal': 'A'}}],
+                                              'comment': acl_comment}},
+                                      {name: {acl_type: [{acl_names[0]:
+                                                         {'octal': 'A'}},
+                                                         {acl_names[1]:
+                                                         {'octal': 'A'}}],
+                                              'comment': acl_comment}},
+                                      {name: {acl_type: [{acl_names[0]:
+                                                         {'octal': 'A'}},
+                                                         {acl_names[1]:
+                                                         {'octal': 'A'}}]}},
+                                      {name: {acl_type: [{}]}},
+                                      {name: {acl_type: [{}]}},
+                                      {name: {acl_type: [{}]}},
+                                      {name: {acl_type: ''}}])
+        mock_modfacl = MagicMock(return_value=True)
+
+        with patch.dict(linux_acl.__salt__, {'acl.getfacl': mock}):
+            # Update - test=True
+            with patch.dict(linux_acl.__opts__, {'test': True}):
+                comt = ('Updated permissions will be applied for {0}: A -> {1}'
+                        ''.format(acl_names, perms))
+                expected = {'name': name,
+                            'comment': comt,
+                            'changes': {},
+                            'pchanges': {'new': {'acl_name': ', '.join(acl_names),
+                                                 'acl_type': acl_type,
+                                                 'perms': 7},
+                                         'old': {'acl_name': ', '.join(acl_names),
+                                                 'acl_type': acl_type,
+                                                 'perms': 'A'}},
+                            'result': None}
+
+                ret = linux_acl.list_present(name, acl_type, acl_names, perms)
+                self.assertDictEqual(ret, expected)
+
+            # Update - test=False
+            with patch.dict(linux_acl.__salt__, {'acl.modfacl': mock_modfacl}):
+                with patch.dict(linux_acl.__opts__, {'test': False}):
+                    comt = ('Applied new permissions for {0}'.format(', '.join(acl_names)))
+                    expected = {'name': name,
+                                'comment': comt,
+                                'changes': {'new': {'acl_name': ', '.join(acl_names),
+                                                    'acl_type': acl_type,
+                                                    'perms': 'rwx'}},
+                                'pchanges': {},
+                                'result': True}
+
+                    ret = linux_acl.list_present(name, acl_type, acl_names, perms)
+                    self.assertDictEqual(expected, ret)
+
+            # Update - modfacl error
+            with patch.dict(linux_acl.__salt__, {'acl.modfacl': MagicMock(
+                    side_effect=CommandExecutionError('Custom err'))}):
+                with patch.dict(linux_acl.__opts__, {'test': False}):
+                    comt = ('Error updating permissions for {0}: Custom err'
+                            ''.format(acl_names))
+                    expected = {'name': name,
+                                'comment': comt,
+                                'changes': {},
+                                'pchanges': {},
+                                'result': False}
+
+                    ret = linux_acl.list_present(name, acl_type, acl_names, perms)
+                    self.assertDictEqual(expected, ret)
+
+            # New - test=True
+            with patch.dict(linux_acl.__salt__, {'acl.modfacl': mock_modfacl}):
+                with patch.dict(linux_acl.__opts__, {'test': True}):
+                    comt = ('New permissions will be applied '
+                            'for {0}: {1}'.format(acl_names, perms))
+                    expected = {'name': name,
+                                'comment': comt,
+                                'changes': {},
+                                'pchanges': {'new': {'acl_name': ', '.join(acl_names),
+                                                     'acl_type': acl_type,
+                                                     'perms': perms}},
+                                'result': None}
+
+                    ret = linux_acl.list_present(name, acl_type, acl_names, perms)
+                    self.assertDictEqual(expected, ret)
+
+            # New - test=False
+            with patch.dict(linux_acl.__salt__, {'acl.modfacl': mock_modfacl}):
+                with patch.dict(linux_acl.__opts__, {'test': False}):
+                    comt = ('Applied new permissions for {0}'.format(', '.join(acl_names)))
+                    expected = {'name': name,
+                                'comment': comt,
+                                'changes': {'new': {'acl_name': ', '.join(acl_names),
+                                                    'acl_type': acl_type,
+                                                    'perms': perms}},
+                                'pchanges': {},
+                                'result': True}
+                    ret = linux_acl.list_present(name, acl_type, acl_names, perms)
+                    self.assertDictEqual(expected, ret)
+
+            # New - modfacl error
+            with patch.dict(linux_acl.__salt__, {'acl.modfacl': MagicMock(
+                    side_effect=CommandExecutionError('Custom err'))}):
+                with patch.dict(linux_acl.__opts__, {'test': False}):
+                    comt = ('Error updating permissions for {0}: Custom err'
+                            ''.format(acl_names))
+                    expected = {'name': name,
+                                'comment': comt,
+                                'changes': {},
+                                'pchanges': {},
+                                'result': False}
+
+                    ret = linux_acl.list_present(name, acl_type, acl_names, perms)
+                    self.assertDictEqual(expected, ret)
+
+            # No acl type
+            comt = ('ACL Type does not exist')
+            expected = {'name': name, 'comment': comt, 'result': False,
+                        'changes': {}, 'pchanges': {}}
+            ret = linux_acl.list_present(name, acl_type, acl_names, perms)
+            self.assertDictEqual(expected, ret)
+
+    # 'list_absent' function tests: 1
+
+    def test_list_absent(self):
+        '''
+        Test to ensure a Linux ACL does not exist
+        '''
+        name = '/root'
+        acl_type = 'users'
+        acl_names = ['damian', 'homer']
+        perms = 'rwx'
+
+        ret = {'name': name,
+               'result': None,
+               'comment': '',
+               'changes': {}}
+
+        mock = MagicMock(side_effect=[{name: {acl_type: [{acl_names[0]: {'octal': 'A'},
+                                                          acl_names[1]: {'octal': 'A'}}]}},
+                                      {name: {acl_type: ''}}])
+        with patch.dict(linux_acl.__salt__, {'acl.getfacl': mock}):
+            with patch.dict(linux_acl.__opts__, {'test': True}):
+                comt = ('Removing permissions')
+                ret.update({'comment': comt})
+                self.assertDictEqual(linux_acl.absent(name, acl_type, acl_names, perms), ret)
+
+            comt = ('ACL Type does not exist')
+            ret.update({'comment': comt, 'result': False})
+            self.assertDictEqual(linux_acl.list_absent(name, acl_type, acl_names), ret)
```

**Comment by robert8103** [comment]
hi @garethgreenaway, awesome, i've a quite hard month! 
Gread thank's for your help! is there some more to do for me? I'm at the moment a bit lost.

**Comment by garethgreenaway** [comment]
@robert8103 Apologies for the delay.  This should be good to go.  I've added the tests to your branch.

**Comment by rallytime** [comment]
The following tests are failing related to this PR:

- unit.states.test_linux_acl.LinuxAclTestCase.test_list_absent
- unit.states.test_linux_acl.LinuxAclTestCase.test_list_present

https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu14-n/23152/

**Pull request merged**